src/main/java/jenkins/plugins/openstack/compute/SlaveOptionsDescriptor.java

```diff
@@ -437,14 +546,68 @@ public FormValidation doCheckSecurityGroups(
     }
 
     @Restricted(DoNotUse.class)
+    @InjectOsAuth
+    public ComboBoxModel doFillAvailabilityZoneItems(
+            @QueryParameter String availabilityZone,
+            @QueryParameter String endPointUrl, @QueryParameter String identity, @QueryParameter String credential, @QueryParameter String zone
+    ) {
+        // Support for availabilityZones is optional in OpenStack, so this is a f:combobox not f:select field.
+        // Therefore we suggest some options if we can, but if we can't then we assume it's because they're not needed.
+        final ComboBoxModel m = new ComboBoxModel();
+        try {
+            if (haveServerDetails(endPointUrl, identity, credential, zone)) {
+                final Openstack openstack = Openstack.Factory.get(endPointUrl, identity, credential, zone);
+                for (final AvailabilityZone az : openstack.getAvailabilityZones()) {
+                    final String value = az.getZoneName();
+                    m.add(value);
+                }
+            }
+        } catch (AuthenticationException | FormValidation | ConnectionException ex) {
+            LOGGER.log(Level.FINEST, "Openstack call failed", ex);
+        } catch (Exception ex) {
+            LOGGER.log(Level.SEVERE, ex.getMessage(), ex);
+        }
+        return m;
+    }
+
+    @Restricted(DoNotUse.class)
     public FormValidation doCheckAvailabilityZone(
             @QueryParameter String value,
-            @RelativePath("../../slaveOptions") @QueryParameter("availabilityZone") String def
-    ) {
+            @RelativePath("../../slaveOptions") @QueryParameter("availabilityZone") String def,
```

Comment:
Ick - the more I look at this, the more I remember just what a nightmare it was to get this working...

The Jenkins code that binds the jelly to the descriptor uses the @QueryParameter as a unique key for passing data back to the Jenkins code BUT it will read from as many @RelativePath locations as you tell it to.  So you can ask for the same named parameter from multiple places, but you will (at most) only get one value (but you'll be passed that multiple times).
So, in this case, we're naming both places that the authentication information might be (relative to this field) so that, where ever it is, we will find it.
Using @InjectAuth would have been much neater, but that doesn't work on doCheck methods, only on doFill methods.